// mongodb-core/index.js
'use strict';

var BSON = require('mongodb-core/node_modules/bson');
var require_optional = require('mongodb-core/node_modules/require_optional');
const EJSON = require('mongodb-core/lib/utils').retrieveEJSON();

try {
  // Attempt to grab the native BSON parser
  var BSONNative = require_optional('bson-ext');
  // If we got the native parser, use it instead of the
  // Javascript one
  if (BSONNative) {
    BSON = BSONNative;
  }
} catch (err) {} // eslint-disable-line

module.exports = {
  // Errors
  MongoError: require('mongodb-core/lib/error').MongoError,
  MongoNetworkError: require('mongodb-core/lib/error').MongoNetworkError,
  MongoParseError: require('mongodb-core/lib/error').MongoParseError,
  MongoTimeoutError: require('mongodb-core/lib/error').MongoTimeoutError,
  MongoWriteConcernError: require('mongodb-core/lib/error').MongoWriteConcernError,
  mongoErrorContextSymbol: require('mongodb-core/lib/error').mongoErrorContextSymbol,
  // Core
  Connection: require('mongodb-core/lib/connection/connection'),
  Server: require('mongodb-core/lib/topologies/server'),
  ReplSet: require('mongodb-core/lib/topologies/replset'),
  Mongos: require('mongodb-core/lib/topologies/mongos'),
  Logger: require('mongodb-core/lib/connection/logger'),
  Cursor: require('mongodb-core/lib/cursor'),
  ReadPreference: require('mongodb-core/lib/topologies/read_preference'),
  Sessions: require('mongodb-core/lib/sessions'),
  BSON: BSON,
  EJSON: EJSON,
  // Raw operations
  Query: require('mongodb-core/lib/connection/commands').Query,
  // Auth mechanisms
  defaultAuthProviders: require('mongodb-core/lib/auth/defaultAuthProviders').defaultAuthProviders,
  MongoCR: require('mongodb-core/lib/auth/mongocr'),
  X509: require('mongodb-core/lib/auth/x509'),
  Plain: require('mongodb-core/lib/auth/plain'),
  GSSAPI: require('mongodb-core/lib/auth/gssapi'),
  ScramSHA1: require('mongodb-core/lib/auth/scram').ScramSHA1,
  ScramSHA256: require('mongodb-core/lib/auth/scram').ScramSHA256,
  // Utilities
  parseConnectionString: require('mongodb-core/lib/uri_parser')
};
