// @umijs/preset-built-in/lib/plugins/features/analyze.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

function _react() {
  const data = _interopRequireDefault(require("react"));

  _react = function _react() {
    return data;
  };

  return data;
}

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var _default = api => {
  api.describe({
    key: 'analyze',
    config: {
      schema(joi) {
        return joi.object({
          analyzerMode: joi.string().valid('server', 'static', 'disabled'),
          analyzerHost: joi.string(),
          analyzerPort: joi.alternatives(joi.number(), 'auto'),
          openAnalyzer: joi.boolean(),
          generateStatsFile: joi.boolean(),
          statsFilename: joi.string(),
          logLevel: joi.string().valid('info', 'warn', 'error', 'silent'),
          defaultSizes: joi.string().valid('stat', 'parsed', 'gzip')
        }).unknown(true);
      },

      default: {
        analyzerMode: process.env.ANALYZE_MODE || 'server',
        analyzerPort: process.env.ANALYZE_PORT || 8888,
        openAnalyzer: process.env.ANALYZE_OPEN !== 'none',
        // generate stats file while ANALYZE_DUMP exist
        generateStatsFile: !!process.env.ANALYZE_DUMP,
        statsFilename: process.env.ANALYZE_DUMP || 'stats.json',
        logLevel: process.env.ANALYZE_LOG_LEVEL || 'info',
        defaultSizes: 'parsed' // stat  // gzip

      }
    },
    enableBy: () => {
      var _api$config;

      const ssr = (_api$config = api.config) === null || _api$config === void 0 ? void 0 : _api$config.ssr;
      return !!(process.env.ANALYZE && !ssr || process.env.ANALYZE_SSR && ssr);
    }
  });
  api.chainWebpack(webpackConfig => {
    var _api$config2;

    webpackConfig.plugin('bundle-analyzer').use(require('umi-webpack-bundle-analyzer').BundleAnalyzerPlugin, [((_api$config2 = api.config) === null || _api$config2 === void 0 ? void 0 : _api$config2.analyze) || {}]);
    return webpackConfig;
  });
};

exports.default = _default;
