// @umijs/preset-built-in/lib/plugins/features/forkTSChecker.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

function _react() {
  const data = _interopRequireDefault(require("react"));

  _react = function _react() {
    return data;
  };

  return data;
}

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var _default = api => {
  api.describe({
    key: 'forkTSCheker',
    config: {
      schema(joi) {
        return joi.object();
      }

    },
    enableBy: () => {
      var _api$config;

      return process.env.FORK_TS_CHECKER || !!((_api$config = api.config) === null || _api$config === void 0 ? void 0 : _api$config.forkTSCheker);
    }
  });
  api.chainWebpack(webpackConfig => {
    webpackConfig.plugin('fork-ts-checker').use(require('fork-ts-checker-webpack-plugin'), [{
      formatter: 'codeframe',
      // parallel
      async: false,
      checkSyntacticErrors: true
    }]);
    return webpackConfig;
  });
};

exports.default = _default;
